intrusive_collections_examples/src/basic_datastructures/singly_linked_list.rs

```diff
@@ -0,0 +1,218 @@
+pub struct SinglyLinkedListNode {
+    /// The pointer to the next node in the linked list.  When this is `Some`,
+    /// there is another pointer to follow, when this is `None`, there is no
+    /// next pointer to follow.
+    next: Option<*const SinglyLinkedListNode>,
+
+    /// Since we're using raw pointers, we **cannot** allow these instances to
+    /// be moved by the compiler (dropping is OK, but moving will almost
+    /// certainly cause UB).  For more information on why this is necessary,
+    /// start reading at
+    /// https://doc.rust-lang.org/std/pin/index.html#example-self-referential-struct
+    /// and be extra careful to read all the way through the explanations about
+    /// drop!
+    _pin: PhantomPinned,
+}
+
+impl SinglyLinkedListNode {
+    /// Sets the next pointer to `None`.
+    ///
+    /// This method is only useful in a few rare circumstances, but when you
+    /// need it, you *really need it.*  The main use is when you're working on
+    /// the head node of a list; you can't extract it, which means that without
+    /// this method, it is impossible to remove it from the list it is a part
+    /// of.
+    ///
+    /// ## Safety
+    ///
+    /// By the rules that govern rust, this method is safe.  **HOWEVER,** it is
+    /// potentially powerful foot-gun.  If the only way you can access the
+    /// elements that follow `self` is via `self.next`, then calling this method
+    /// will mean that you no longer have access to the rest of the list.  Make
+    /// sure that you are already holding the next object *before* you null out
+    /// the next pointer!
+    pub fn set_next_to_null(&mut self) {
+        self.next = None;
+    }
+
+    /// Inserts `self` immediately after `previous`.
+    ///
+    /// If this method successfully completes, then `self` will be inserted into
+    /// the linked list immediately after previous.  If there are detectable
+    /// errors, then this will panic on the error.
+    ///
+    /// ## Parameters
+    ///
+    /// - `previous` - This is the node that will become `self`'s prior node. If
+    ///   it is already pointing to something, then `self` will end up pointing
+    ///   to that object, while `previous` will point to self.
+    ///
+    /// ## Safety
+    ///
+    /// This method is unsafe.  There is no proof that `previous` points to a
+    /// valid memory location, it's just an assumption that we're forced to
+    /// make. In addition, if `self` is already pointing to something, then that
+    /// something will be dropped silently.  Finally, we assume that our singly
+    /// linked list is not a circular list.  This method will permit you to
+    /// create a circularly linked list, including make a node that links to
+    /// itself.
+    pub unsafe fn insert(&mut self, previous: *mut SinglyLinkedListNode) {
+        self.next = (*previous).next;
+        let me: *const SinglyLinkedListNode = self;
```

Comment:
You can't use `&mut self` here, because `SinglyLinkedListNode` is intrinsically shared. (By the node it's in, and by the previous node). So you have to use `&self` or raw pointers. I think you can't use `&self` because that would prevent going from `self.next` -> node, but check with MIRI after you complete your implementation.

I made some notes about this in my previous review: https://github.com/RustyYato/generic-field-projection/issues/47#issuecomment-754940068

Reply:
Good catch, thank you!